### PR TITLE
Introduce `types.anything`

### PR DESCRIPTION
**Pull request by infinisil**
###### Motivation for this change
This introduces the new type `types.anything`, which implements a safe and reasonable merge strategy if you don't care about the type. Specifically:
- It recursively merges attribute sets together
  - Except if it's string-coercible (aka, packages and co.), then only one definition is allowed
- It doesn't concatenate list values together (since list concatenation doesn't always make sense)
- If it's an atomic value (not a set or a list), all definitions must be equal
- It applies `mkIf`'s and co. for all values

This is in contrast with the often-used `types.attrs` and `types.unspecified`, which can be problematic because they don't adhere to above points. If both this PR and https://github.com/NixOS/nixpkgs/pull/97114 are merged, I can start deprecating those in favor of this one.

Note that the doc entry for this new type is intentionally kept small, because its behavior is very natural and doesn't need a special explanation.

Ping @rycee @roberth @Profpatsch

###### Things done

- [x] Wrote a doc entry for the new type
- [x] Wrote tests for the new type

### CODE REVIEW

**Comment by edolstra** [comment]
Hm, what's the use case for this? Presumably you always *do* care about the type of an option since you need to use it somewhere eventually.

**Comment by rycee** [comment]
@edolstra I don't think the `anything` type is very useful by itself but I would happily use, e.g., `attrsOf anything` instead of `attrs`. That is, when I know the type is an attribute set but not the specific content. I've encountered this mainly when defining `settings`-like options that gets serialized to some hierarchical configuration format. For example the [`programs.taskwarrier.config`](https://github.com/rycee/home-manager/blob/a063b73d5cf2d741f5f873897974d1a4c3e36d3d/modules/programs/taskwarrior.nix#L42-L64) option in Home Manager. (It should really be named `programs.taskwarrier.settings` but it is an old module 🙂).

**Comment by Infinisil** [comment]
Yeah exactly, the intention is for this type to replace `types.attrs` and `types.unspecified`, which are currently used when the type isn't known, but as mentioned they have some surprising behavior which you wouldn't expect.

File: nixos/doc/manual/development/option-types.xml [hunk 1]
```diff
@@ -21,6 +21,16 @@
   </para>
 
   <variablelist>
+   <varlistentry>
+    <term>
+     <varname>types.anything</varname>
+    </term>
+    <listitem>
+     <para>
+      A type that accepts any value and merges attribute sets together.
```
Comment (Profpatsch) [inline nixos/doc/manual/development/option-types.xml:30]: I would give some more explanation and examples here. At least one working and one failing example. Nothing is truly trivial for every reader :)

File: lib/types.nix [hunk 2]
```diff
@@ -101,6 +101,42 @@ rec {
...
+          # don't have the same type
+          commonType = foldl' (type: def:
+            if getType def.value == type
+            then type
+            else throw "The option `${showOption loc}' has conflicting option types in ${showFiles (getFiles defs)}"
+          ) (getType (head defs).value) defs;
+
+          mergeFunction = {
+            # Recursively merge attribute sets
+            set = (attrsOf anything).merge;
```
Comment (Profpatsch) [inline lib/types.nix:126]: Is that a right-based merge?

**Review by Profpatsch — CHANGES_REQUESTED** [review]

File: lib/types.nix (hunk 2 above)
Comment (Infinisil) [reply lib/types.nix:126]: It merges recursively. So it first combines all attributes of the same name together, then merges those with the `anything` merge function, which errors if there's conflicting definitions. So no side has precedence.

File: nixos/doc/manual/development/option-types.xml (hunk 1 above)
Comment (Infinisil) [reply nixos/doc/manual/development/option-types.xml]: Added an example now

**Comment by Infinisil** [comment]
Changes since the last revision:
- I moved `types.attrs` to the bottom of the type list and added a warning about its problems
- Added an example for `types.anything` in the docs
- Fixed `mergeEqualOption` for when a single function is passed. Now it doesn't error out anymore. Also added a test case to `types.anything` for this

**Review by roberth — APPROVED** [review]

**Review by Profpatsch — APPROVED** [review]
LGTM

**Comment by roberth** [comment]
Thanks @Infinisil !

**Pull request merged**